src/llama-model.cpp

```diff
@@ -1753,6 +1760,13 @@ bool llama_model::load_tensors(llama_model_loader & ml) {
                             layer.ffn_gate_exps = create_tensor(tn(LLM_TENSOR_FFN_GATE_EXPS, "weight", i), {n_embd,   n_ff, n_expert}, TENSOR_NOT_REQUIRED);
                             layer.ffn_down_exps = create_tensor(tn(LLM_TENSOR_FFN_DOWN_EXPS, "weight", i), {  n_ff, n_embd, n_expert}, 0);
                             layer.ffn_up_exps   = create_tensor(tn(LLM_TENSOR_FFN_UP_EXPS,   "weight", i), {n_embd,   n_ff, n_expert}, 0);
+
+                            // For Granite MoE Shared
+                            if (arch == LLM_ARCH_GRANITE_MOE_SHARED) {
+                                layer.ffn_gate_shexp = create_tensor(tn(LLM_TENSOR_FFN_GATE_SHEXP, "weight", i), {n_embd, hparams.n_ff_shexp}, 0);
```

Comment:
I think you can simply check `hparams.n_ff_shexp > 0` then load these tensors. The value of `hparams.n_ff_shexp` will be 0 if it's not written in gguf

This way, you can remove 2/3 code of this PR, making it much easier

Reply:
Ah, yes, this would be _much_ simpler! All of the granite architectures are highly related and slight tweaks on other architectures. In `transformers`, the approach has always been to keep the architectures isolated and use their `modular` approach for shared code. I know that policy is less strict here, but wasn't sure how much to lean into collapsing all of these architectures into a single architecture string versus keeping them separate and sharing code underneath. Is there a strong preference here going forward?

Reply:
Makes sense. My personal preference is to keep a 1:1 mapping between architecture names in `huggingface`, then have maximal code reuse on the backend. That is a very weakly held preference, though, so I'm happy to take whatever direction is best from the maintainers' POV.

Reply:
@gabe-l-hart Not sure if you have a near deadline to merge this PR (?)

We never aim for 1:1 mapping between HF `model_type` <> llama.cpp arch name from the beginning, so it's probably fine to consider this new arch as a variant of `GRANITE_MOE` with `n_ff_shexp > 0`. I'm not sure in which case the 1:1 mapping will be useful?

In anw, after second thought, I have no strong opinion whether this need to be a dedicated arch name or not, you can probably keep it this way if you like. But maybe we should move the cgraph of granite family to a dedicated `build_*` function. My main concern is that we're adding things on top of `build_llama` which make it not very easy to read